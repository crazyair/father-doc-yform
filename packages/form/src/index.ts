import { Form as AntdForm } from 'antd';
import { FormContext, FormItemContext } from './Context';
import InternalForm, { config } from './form';
import Items from './items';

import './styles/index.less';

type InternalFormType = typeof InternalForm;

interface FormInterface extends InternalFormType {
  Items: typeof Items;
  config: typeof config;
  useForm: typeof AntdForm.useForm;
  FormContext: typeof FormContext;
  FormItemContext: typeof FormItemContext;
}

const Form = InternalForm as FormInterface;
Form.Items = Items;
Form.config = config;
Form.useForm = AntdForm.useForm;
Form.FormContext = FormContext;
Form.FormItemContext = FormItemContext;

export { Form };
