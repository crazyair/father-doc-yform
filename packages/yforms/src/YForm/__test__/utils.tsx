import { act } from 'react-dom/test-utils';
import { ReactWrapper } from 'enzyme';

export const delay = (timeout = 0) =>
  new Promise((resolve) => {
    setTimeout(resolve, timeout);
  });

export async function waitForComponentToPaint<P = {}>(wrapper: ReactWrapper<P>, amount = 0) {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, amount));
    wrapper.update();
  });
}
