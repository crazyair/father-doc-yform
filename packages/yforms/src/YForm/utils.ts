import { isValidElement, useRef } from 'react';
import { isImmutable } from 'immutable';
import {
  get,
  map,
  join,
  set,
  mapKeys,
  forEach,
  sortBy,
  isArray,
  find,
  isEqual,
  mergeWith,
} from 'lodash';
import { ColProps } from 'antd/lib/col';

import warning from 'warning';
import { stringAndFunc } from './ItemsType';
import { KeyValue, ParamsType } from './Form';
import { FormatFieldsValue, YFormItemProps } from './Items';

const nzhcn = require('nzh/cn');

export const layoutMore = {
  labelCol: { xs: { span: 24 }, sm: { span: 4 } },
  wrapperCol: { xs: { span: 24 }, sm: { span: 20 } },
};
export const tailLayout = { wrapperCol: { offset: 8, span: 16 } };

export function replaceMessage(template: string, kv: Record<string, string>): string {
  return template.replace(/\$\{\w+\}/g, (str: string) => {
    const key = str.slice(2, -1);
    return kv[key];
  });
}

export const mergeWithDom = (obj: any, ...params: any[]) => {
  return mergeWith(obj, ...params, (_, srcValue) => {
    // 如果是元素则返回要更改的值，不是则不处理
    if (isValidElement(srcValue)) {
      return srcValue;
    }
    // 如果是不可变数据，不处理合并
    if (isImmutable(srcValue)) {
      return srcValue;
    }
  });
};

// 获取一行多组件的 width
export const oneLineItemStyle = (list?: Array<number | string>) => {
  if (!list || !Array.isArray(list)) return [];
  const _list: { display: string; width: string }[] = [];
  let width = 0;
  let count = 0;
  list.forEach((item) => {
    if (typeof item === 'number') {
      width += item;
    } else {
      count += 1;
    }
  });

  list.forEach((item) => {
    if (typeof item === 'number') {
      _list.push({ display: 'inline-block', width: `${item}px` });
    } else {
      _list.push({ display: 'inline-block', width: `calc(${item} - ${width / count}px)` });
    }
  });
  return _list;
};

export function getFieldKeyValue<T>(record: T, index: number, field: stringAndFunc<T>) {
  const recordKey = typeof field === 'function' ? field(record, index) : get(record, field);
  return recordKey === undefined ? index : recordKey;
}

export const searchSelect = {
  allowClear: true,
  showSearch: true,
  optionFilterProp: 'children',
  filterOption: (input: string, option: any) => {
    const getValue = (dom: React.ReactNode): string => {
      const _value = get(dom, 'props.children');
      if (Array.isArray(_value)) {
        const d = map(_value, (item) => {
          if (isValidElement(item)) {
            return getValue(item);
          }
          return item;
        });
        return join(d, '');
      }
      if (isValidElement(_value)) {
        return getValue(_value);
      }
      return join(_value, '');
    };

    const str = getValue(option);
    return str.toLowerCase().indexOf(input.toLowerCase()) >= 0;
  },
};

// jiesuan 项目中使用的计算中文、全角字符 x2
export const calculateStrLength = (name?: string | number): number => {
  if (name === null || name === void 0) return 0;
  if (typeof name === 'number') {
    name = `${name}`;
  }
  let count = 0;
  const strArr = Array.from(name);
  strArr.forEach((c) => {
    if (/[\x00-\xff]/.test(c)) {
      count++;
    } else {
      count += 2;
    }
  });
  return count;
};

export const convertMoney = (money: string | number) => {
  return nzhcn.encodeB(money, { tenMin: true });
};

interface NoLabelLayoutValueProps {
  labelCol?: ColProps;
  wrapperCol?: ColProps;
  offset?: number;
}
// 处理 label 宽度
export const getLabelLayout = ({ labelCol, wrapperCol, offset = 0 }: NoLabelLayoutValueProps) => {
  const labelLayoutValue = {};
  const noLabelLayoutValue = {};
  const labelSpan = get(labelCol, 'span');
  const wrapperSpan = get(wrapperCol, 'span');
  if (labelSpan) {
    set(labelLayoutValue, ['labelCol', 'span'], Number(labelSpan) + offset);
    set(labelLayoutValue, ['wrapperCol', 'span'], Number(wrapperSpan) - offset);
    set(noLabelLayoutValue, ['wrapperCol', 'offset'], Number(labelSpan) + offset);
    set(noLabelLayoutValue, ['wrapperCol', 'span'], Number(wrapperSpan) - offset);
  } else {
    mapKeys(labelCol, (value, key) => {
      set(labelLayoutValue, ['labelCol', key, 'span'], value.span + offset);
      set(noLabelLayoutValue, ['wrapperCol', key, 'offset'], value.span + offset);
    });
    mapKeys(wrapperCol, (value, key) => {
      set(labelLayoutValue, ['wrapperCol', key, 'span'], value.span - offset);
      set(noLabelLayoutValue, ['wrapperCol', key, 'span'], value.span - offset);
    });
  }

  return { noLabelLayoutValue, labelLayoutValue };
};
// 返回上一级 name 的数据
export const getParentNameData = (values: any, name: YFormItemProps['name']) => {
  const _values = { ...values };
  const _name = isArray(name) ? name : [name];
  if (_name.length === 1) {
    return _values;
  }
  return get(_values, _name.slice(0, _name.length - 1));
};

export function submitFormatValues(
  values: KeyValue,
  formatFieldsValue?: FormatFieldsValue[],
): KeyValue {
  const _values = mergeWithDom({}, values);
  const list: FormatFieldsValue[] = sortBy(formatFieldsValue, (item) => {
    if (!item) return;
    if (isArray(item.name)) {
      return -item.name.length;
    }
    return -`${item.name}`.length;
  }).filter((x) => x);
  forEach(list, (item) => {
    const { name, format } = item;
    if (name && format) {
      const parentValue = getParentNameData(values, name);
      // 如果上一级是 undefined，则不处理该字段。（List add 会生成空对象）
      if (parentValue === undefined) return;
      try {
        set(_values, name, format(get(values, name), parentValue, values));
      } catch (error) {
        // 如果 format 代码报错这里抛出异常
        // eslint-disable-next-line no-console
        console.error(error);
        warning(false, error);
      }
    }
  });
  return _values;
}

export const onFormatFieldsValue = (formatFieldsValue: FormatFieldsValue[]) => {
  return (list: FormatFieldsValue[]) => {
    const _formatFields = formatFieldsValue;
    forEach(list, (item) => {
      // 已存在不再注册
      if (!find(_formatFields, { name: item.name })) {
        _formatFields.push(item);
      }
    });
    return _formatFields;
  };
};

export const paramsType = (params?: ParamsType) => {
  const _params = params || ({} as ParamsType);
  const type = {
    id: _params.id,
    edit: _params.type === 'edit',
    create: _params.type === 'create',
    view: _params.type === 'view',
  };
  let typeName = '';
  if (type.create) typeName = '新建';
  if (type.edit) typeName = '编辑';
  if (type.view) typeName = '查看';

  return { ...type, typeName };
};

export const useImmutableValue = (value: any) => {
  const v = useRef(value);
  if (!isEqual(value, v.current)) {
    v.current = value;
  }
  return v.current;
};
