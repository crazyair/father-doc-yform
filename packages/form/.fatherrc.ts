import { IBundleOptions } from 'father-build/src/types';

const options: IBundleOptions = {
  // entry: 'src/index.tsx',
  // cjs: 'rollup',
  // esm: 'rollup',
  // target: 'node',
  runtimeHelpers: true,
  cjs: { type: 'babel', lazy: true },
  esm: 'babel',
  cssModules: false, // https://github.com/umijs/father/issues/131
  // extractCSS: true,
  // lessInBabelMode: true,
  extraBabelPlugins: [
    ['babel-plugin-import', { libraryName: 'antd', libraryDirectory: 'es', style: true }],
  ],
};

export default options;
