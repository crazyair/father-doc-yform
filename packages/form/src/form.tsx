import React from 'react';
import { Form as AntdForm } from 'antd';
import { FormInstance, FormProps as AntdFormProps } from 'antd/lib/form';
import Items from './items';
import { FormItemsType } from './ItemsType';

export type childrenType = FormItemsType | FormItemsType[];

export interface FormProps<Values = any> extends Omit<AntdFormProps<Values>, 'children'> {
  children?: childrenType | childrenType[];
}

const InternalForm: React.ForwardRefRenderFunction<unknown, FormProps> = (props, ref) => {
  const { children, form, ...rest } = props;

  const [wrapForm] = AntdForm.useForm(form);

  React.useImperativeHandle(ref, () => wrapForm);

  return (
    <AntdForm form={wrapForm} {...rest}>
      <Items>{children}</Items>
    </AntdForm>
  );
};

const Form = React.forwardRef<FormInstance, FormProps>(InternalForm) as <Values = any>(
  props: FormProps<Values> & { ref?: React.Ref<FormInstance<Values>> },
) => React.ReactElement;

export default Form;
