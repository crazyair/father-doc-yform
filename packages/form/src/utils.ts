import { find, forEach, get, isArray, isObject, set } from 'lodash';
import React from 'react';
import { FormatFieldsValue, FormProps, ItemsType } from './form';

// 获取唯一 key
export const getOnlyKey = () => {
  const keyMap = new Set();
  return (index, pIndex, name) => {
    let _index = name ? name.toString() : index;
    let key = pIndex ? `${pIndex}_${_index}` : _index;
    if (keyMap.has(key)) {
      _index = `${_index}_${index}`;
      key = pIndex ? `${pIndex}_${_index}` : _index;
    }
    keyMap.add(key);
    return key;
  };
};

export const onFormatFieldsValue = (formatFieldsValue: FormatFieldsValue[]) => {
  return (list: FormatFieldsValue[]) => {
    const _formatFields = formatFieldsValue;
    forEach(list, (item) => {
      // 已存在不再注册
      if (!find(_formatFields, { name: item.name })) {
        _formatFields.push(item);
      }
    });
    return _formatFields;
  };
};

export const getChildrenTypes = (children: FormProps['children'], initialValues = {}) => {
  const obj = {};
  const each = (children: FormProps['children'], initialValues = {}) => {
    forEach(isArray(children) ? children : [children], (item) => {
      if (isArray(item)) {
        return each(item, initialValues);
      }
      if (React.isValidElement(item)) {
        return;
      }
      if (isObject(item)) {
        const { name, deFormat } = item as ItemsType;
        if (deFormat) {
          set(obj, name, deFormat(get(initialValues, name), initialValues));
        }
      }
    });
    return obj;
  };
  each(children, initialValues);
  return obj;
};
