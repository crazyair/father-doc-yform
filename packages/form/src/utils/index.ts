import React from 'react';
import { forEach, get, isArray, merge, mergeWith, set } from 'lodash';
import { isImmutable } from 'immutable';
import { FormItemProps } from 'antd/lib/form';

import { FormatFieldsValue } from '../form';

// 返回上一级 name 的数据
export const getParentNameData = (values: any, name: FormItemProps['name']) => {
  const _values = { ...values };
  const _name = isArray(name) ? name : [name];
  if (_name.length === 1) {
    return _values;
  }
  return get(_values, _name.slice(0, _name.length - 1));
};

export function submitFormatValues(values: any, formatFieldsValue?: FormatFieldsValue[]) {
  const _values = merge({}, values);
  forEach(formatFieldsValue, (item) => {
    const { name, format } = item;
    const parentValue = getParentNameData(values, name);
    // 如果上一级是 undefined，则不处理该字段。（List add 会生成空对象）
    if (parentValue === undefined) return;
    if (name && format) {
      set(_values, name, format(get(values, name), values));
    }
  });
  return _values;
}

// 获取一行多组件的 width
export const oneLineItemStyle = (list?: (number | string)[]) => {
  if (!list || !Array.isArray(list)) return [];
  const _list: { display: string; width: string }[] = [];
  let width = 0;
  let count = 0;
  list.forEach((item) => {
    if (typeof item === 'number') {
      width += item;
    } else {
      count += 1;
    }
  });

  list.forEach((item) => {
    if (typeof item === 'number') {
      _list.push({ display: 'inline-block', width: `${item}px` });
    } else {
      _list.push({ display: 'inline-block', width: `calc(${item} - ${width / count}px)` });
    }
  });
  return _list;
};

export type stringAndFunc<T> = string | ((record: T, index: number) => React.ReactNode);

export function getFieldKeyValue<T>(record: T, index: number, field: stringAndFunc<T>) {
  const recordKey = typeof field === 'function' ? field(record, index) : get(record, field);
  return recordKey === undefined ? index : recordKey;
}

export const mergeWithDom = (obj: any, ...params: any[]) => {
  return mergeWith(obj, ...params, (_, srcValue) => {
    // 如果是元素则返回要更改的值，不是则不处理
    if (React.isValidElement(srcValue)) {
      return srcValue;
    }
    // 如果是不可变数据，不处理合并
    if (isImmutable(srcValue)) {
      return srcValue;
    }
  });
};

export function replaceMessage(template: string, kv: Record<string, string>): string {
  return template.replace(/\$\{\w+\}/g, (str: string) => {
    const key = str.slice(2, -1);
    return kv[key];
  });
}
