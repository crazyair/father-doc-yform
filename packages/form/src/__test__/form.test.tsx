import React from 'react';
import { render } from '@testing-library/react';
import { mount } from 'enzyme';
import { FormInstance } from 'antd/lib/form';
import { Input } from 'antd';
import { Form } from '..';
import { sleep } from '../../tests/utils';

describe('form', () => {
  test('renders', () => {
    const wrapper = render(
      <Form plugins={{ placeholder: { enable: true }, required: { enable: true } }}>
        {[
          { type: 'input', name: 'age', label: 'age' },
          { type: 'radio', name: 'radio', rules: [{ required: true }] },
          {
            type: 'radio',
            name: 'radio2',
            componentProps: {
              showField: 'id',
              postField: 'name',
              options: [{ id: '1', name: 'age' }],
            },
          },
          {
            type: 'radio',
            name: 'radio2',
            componentProps: {
              showField: (item) => item.id,
              postField: () => undefined,
              options: [{ id: '1', name: 'age' }],
            },
          },
          { type: 'custom', name: 'custom', component: <Input /> },
        ]}
      </Form>,
    );
    expect(wrapper).toMatchSnapshot();
  });
  test('config has placeholder', async () => {
    const wrapper = mount(
      <Form plugins={{ placeholder: { enable: true } }}>
        {[{ type: 'input', name: 'age', label: 'age' }]}
      </Form>,
    );
    expect(wrapper.find(Input).at(0).props().placeholder).toBe('请输入age');
  });
  test('init config has placeholder', async () => {
    Form.config({ plugins: { placeholder: { enable: true } } });
    const wrapper = mount(<Form>{[{ type: 'input', name: 'age', label: 'age' }]}</Form>);
    expect(wrapper.find(Input).at(0).props().placeholder).toBe('请输入age');
  });
  test('initFormat format', async () => {
    const ref = React.createRef<FormInstance>();
    const onFinish = jest.fn().mockImplementation(() => {});

    const wrapper = mount(
      <Form
        ref={ref}
        onFinish={(v) => {
          onFinish(v);
        }}
      >
        {[
          { type: 'input', initFormat: () => 1, format: () => 2, name: 'age' },
          { type: 'input', format: () => 2, name: ['name', 'age'] },
          { type: 'input', name: 'remove', format: [{ name: 'remove', removeField: true }] },
        ]}
      </Form>,
    );
    wrapper.find('form').simulate('submit');
    await sleep();
    expect(ref.current.getFieldsValue(true).age).toBe(1);
    expect(onFinish).toHaveBeenLastCalledWith({ age: 2, name: { age: 2 } });
  });
  test('loading', async () => {
    const ref = React.createRef<FormInstance>();

    mount(
      <Form ref={ref} loading initialValues={{ age: 1 }}>
        {[{ type: 'input', name: 'age' }]}
      </Form>,
    );
    expect(ref.current.getFieldsValue(true).age).toBe(undefined);
  });
});
