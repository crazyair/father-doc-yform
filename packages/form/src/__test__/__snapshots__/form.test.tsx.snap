// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-horizontal"
      >
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="age"
              title="age"
            >
              age
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="age"
                  placeholder="请输入age"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-radio-group ant-radio-group-outline"
                  id="radio"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-radio-group ant-radio-group-outline"
                  id="radio2"
                >
                  <label
                    class="ant-radio-wrapper"
                  >
                    <span
                      class="ant-radio"
                    >
                      <input
                        class="ant-radio-input"
                        type="radio"
                        value="age"
                      />
                      <span
                        class="ant-radio-inner"
                      />
                    </span>
                    <span>
                      1
                    </span>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-radio-group ant-radio-group-outline"
                  id="radio2"
                >
                  <label
                    class="ant-radio-wrapper"
                  >
                    <span
                      class="ant-radio"
                    >
                      <input
                        class="ant-radio-input"
                        type="radio"
                        value="0"
                      />
                      <span
                        class="ant-radio-inner"
                      />
                    </span>
                    <span>
                      1
                    </span>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input"
                  id="custom"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="age"
            title="age"
          >
            age
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="age"
                placeholder="请输入age"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-radio-group ant-radio-group-outline"
                id="radio"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-radio-group ant-radio-group-outline"
                id="radio2"
              >
                <label
                  class="ant-radio-wrapper"
                >
                  <span
                    class="ant-radio"
                  >
                    <input
                      class="ant-radio-input"
                      type="radio"
                      value="age"
                    />
                    <span
                      class="ant-radio-inner"
                    />
                  </span>
                  <span>
                    1
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-radio-group ant-radio-group-outline"
                id="radio2"
              >
                <label
                  class="ant-radio-wrapper"
                >
                  <span
                    class="ant-radio"
                  >
                    <input
                      class="ant-radio-input"
                      type="radio"
                      value="0"
                    />
                    <span
                      class="ant-radio-inner"
                    />
                  </span>
                  <span>
                    1
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="custom"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
